src/firmware_main_loop.c

```diff
@@ -30,28 +35,66 @@ void firmware_main_loop(void)
     // This starts the async orientation screen workflow, which is processed by the loop below.
     orientation_screen();
 
+    const uint8_t* hww_data = NULL;
+    uint8_t hww_frame[USB_REPORT_SIZE] = {0};
+
+#if APP_U2F == 1
+    const uint8_t* u2f_data = NULL;
+    uint8_t u2f_frame[USB_REPORT_SIZE] = {0};
+#endif
+
     while (1) {
+        // Do USB I/O
+        if (hid_hww_read(&hww_frame[0])) {
+            util_log("usb: rx data");
+            usb_packet_process((const USB_FRAME*)hww_frame);
+        }
+#if APP_U2F == 1
+        if (hid_u2f_read(&u2f_frame[0])) {
+            util_log("u2f: rx data");
+            u2f_packet_process((const USB_FRAME*)u2f_frame);
+        }
+#endif
+        if (!hww_data) {
+            hww_data = queue_pull(queue_hww_queue());
+        }
+        if (hww_data) {
+            if (hid_hww_write_poll(hww_data)) {
+                util_log("usb: tx data");
+                hww_data = NULL;
+            }
+        }
+#if APP_U2F == 1
+        if (!u2f_data) {
+            u2f_data = queue_pull(queue_u2f_queue());
+        }
+        if (u2f_data) {
+            if (hid_u2f_write_poll(u2f_data)) {
+                util_log("u2f: tx data");
+                u2f_data = NULL;
+            }
+        }
+#endif
+
         screen_process();
         /* And finally, run the high-level event processing. */
 
         rust_workflow_spin();
 
-        if (usb_is_enabled()) {
```

Comment:
Why is this check gone?

Reply:
`hid_usb_read` will return false if usb is disabled.